### ISSUE TITLE
avoid OOM when call _forward() & logprobs

### ISSUE THREAD

**Issue by shiwanghua** [huggingface/trl#3701]
### Feature request

https://github.com/huggingface/trl/blob/d98d53983b6cf6b3381cb084c75b93d8e1ba4c52/trl/trainer/online_dpo_trainer.py#L508

This line is easy to lead to OOM when prompt is very long and max_token is big such as 8192 on 2 A100-40GB.

So we can let cursor to rewrite it, calculate the logit one by one:

```
    def _forward(self, model, prompt_ids, prompt_mask, completion_ids, completion_mask):
        """
        修复切片问题的版本
        """
        batch_size = prompt_ids.size(0)
        all_logprobs = []
        
        for i in range(batch_size):
            # 获取单个样本
            single_prompt_ids = prompt_ids[i:i+1]
            single_prompt_mask = prompt_mask[i:i+1]
            single_completion_ids = completion_ids[i:i+1]
            single_completion_mask = completion_mask[i:i+1]
            
            # 计算截断
            num_tokens_to_truncate = max(
                single_prompt_ids.size(1) + single_completion_ids.size(1) - self.max_length, 
                0
            )
            
            # 截断prompt
            single_prompt_ids = single_prompt_ids[:, num_tokens_to_truncate:]
            single_prompt_mask = single_prompt_mask[:, num_tokens_to_truncate:]
            
            # 拼接
            prompt_completion_ids = torch.cat((single_prompt_ids, single_completion_ids), dim=1)
            prompt_completion_mask = torch.cat((single_prompt_mask, single_completion_mask), dim=1)
            
            # 推理
            with torch.no_grad():
                output = model(prompt_completion_ids, attention_mask=prompt_completion_mask)
            print(f'prompt_completion_ids.shape={prompt_completion_ids.shape}, prompt_completion_mask.shape={prompt_completion_mask.shape}')
            
            # 修复：确保切片索引有效
            prompt_length = single_prompt_ids.size(1)
            if prompt_length > 0:
                # 获取completion部分的logits
                logits = output.logits[:, prompt_length - 1 : -1]
            else:
                # 如果prompt被完全截断，直接使用所有logits
                logits = output.logits
            
            # 确保logits和completion_ids的维度匹配
            if logits.size(1) != single_completion_ids.size(1):
                # 如果长度不匹配，截断到较短的长度
                min_length = min(logits.size(1), single_completion_ids.size(1))
                logits = logits[:, :min_length]
                single_completion_ids = single_completion_ids[:, :min_length]
            
            # 计算logprobs
            logprobs = torch.take_along_dim(
                logits.log_softmax(dim=-1), 
                single_completion_ids.unsqueeze(-1), 
                dim=2
            ).squeeze(-1)
            
            all_logprobs.append(logprobs)
            
            # 清理缓存
            torch.cuda.empty_cache()
        
        return torch.cat(all_logprobs, dim=0)
```

### Motivation

avoid OOM

### Your contribution

rewrite _forward() function

**Comment by shirinyamani** [comment]
Hi @shiwanghua, i wanna understand more abt what you mean. 

is my understanding correct, that imagine we have a batch which is already too long, you wanna break it down to small chunks, each of which is the prompt and forward them then off to next prompts?

is this understanding correct? 

so one long prompt ---> generate one big completion ----> forward

**Comment by shiwanghua** [comment]
> Hi [@shiwanghua](https://github.com/shiwanghua), i wanna understand more abt what you mean.
> 
> is my understanding correct, that imagine we have a batch which is already too long, you wanna break it down to small chunks, each of which is the prompt and forward them then off to next prompts?
> 
> is this understanding correct?
> 
> so one long prompt ---> generate one big completion ----> forward

yes, here the `batch_size=2`

https://github.com/huggingface/trl/blob/68db24e01051d9990cfb93bf7c8abc896462513a/trl/trainer/online_dpo_trainer.py#L479

**Comment by shirinyamani** [comment]
@shiwanghua  ok, im wondering what is the difference of this approach vs grad_accum? 
if you decrease and bacth_size n increase grad_accum you wont get it running?

**Comment by shiwanghua** [comment]
> [@shiwanghua](https://github.com/shiwanghua) ok, im wondering what is the difference of this approach vs grad_accum? if you decrease and bacth_size n increase grad_accum you wont get it running?

There exists bugs in this forward() function, I hope the effect of this function is equivalent to the previous function, but there was an error reported：`element 0 of tensors does not require grad and does not have a grad_fn` when call backward()

https://github.com/huggingface/trl/blob/68db24e01051d9990cfb93bf7c8abc896462513a/trl/trainer/online_dpo_trainer.py#L693

**Comment by shiwanghua** [comment]
remove the `with torch.no_grad():` codes, then I trained successfully.

**Comment by shiwanghua** [comment]
The finnal version:

```
def _forward(self, model, prompt_ids, prompt_mask, completion_ids, completion_mask):
        batch_size = prompt_ids.size(0)
        all_logprobs = []
        for i in range(batch_size):
            # 获取单个样本
            single_prompt_ids = prompt_ids[i:i+1]
            single_prompt_mask = prompt_mask[i:i+1]
            single_completion_ids = completion_ids[i:i+1]
            single_completion_mask = completion_mask[i:i+1]
            
            # 计算截断
            num_tokens_to_truncate = max(single_prompt_ids.size(1) + single_completion_ids.size(1) - self.max_length, 0)
            
            # 截断prompt
            single_prompt_ids = single_prompt_ids[:, num_tokens_to_truncate:]
            single_prompt_mask = single_prompt_mask[:, num_tokens_to_truncate:]
            
            # 如果prompt被完全截断，进一步截断completion_ids
            if single_prompt_ids.size(1) == 0:
                if single_completion_ids.size(1) > self.max_length :
                    single_completion_ids = single_completion_ids[:, -self.max_length:]
                    single_completion_mask = single_completion_mask[:, -self.max_length:]
                    print(f"截断completion_ids到长度: {self.max_length }")
                    
            # 拼接
            prompt_completion_ids = torch.cat((single_prompt_ids, single_completion_ids), dim=1) # [1, 1235]
            prompt_completion_mask = torch.cat((single_prompt_mask, single_completion_mask), dim=1) # [1, 1235]
            
            # 推理
            output = model(prompt_completion_ids, attention_mask=prompt_completion_mask)
            # 修复：确保切片索引有效
            prompt_length = single_prompt_ids.size(1)
            if prompt_length > 0:
                # 获取completion部分的logits
                logits = output.logits[:, prompt_length - 1 : -1]
            else:
                # 如果prompt被完全截断，直接使用所有logits
                logits = output.logits
            # 确保logits和completion_ids的维度匹配
            if logits.size(1) != single_completion_ids.size(1):
                # 如果长度不匹配，截断到较短的长度
                min_length = min(logits.size(1), single_completion_ids.size(1))
                logits = logits[:, :min_length]
                single_completion_ids = single_completion_ids[:, :min_length]

            logprobs = torch.take_along_dim(
                logits.log_softmax(dim=-1), 
                single_completion_ids.unsqueeze(-1), 
                dim=2
            ).squeeze(-1)

            all_logprobs.append(logprobs)

        return torch.cat(all_logprobs, dim=0)
```

**Comment by qgallouedec** [comment]
Hey @shiwanghua, sorry for the very slow reply here, we're catching up on the backlog.

The batched forward you're hitting is the standard memory/throughput trade-off, and TRL already exposes the right knob for it: set `per_device_train_batch_size=1` and raise `gradient_accumulation_steps` to keep your effective batch. That gives `_forward` the same one-sequence-at-a-time memory profile as your loop, but keeps gradients flowing and doesn't serialize the batched path for everyone else, which is also why your first attempt only trained once you dropped the `torch.no_grad()`.

Rewriting `_forward` into a per-sample Python loop (with per-sample truncation and in-loop `empty_cache()`) would slow every Online DPO run for a case the batch-size knob already covers.
